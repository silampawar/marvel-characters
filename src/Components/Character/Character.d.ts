export interface ICharacterDataWrapper {
    code: number;
    status: string;
    copyright: string;
    attributionText: string;
    attributionHTML: string;
    data: ICharacterDataContainer;
    etag: string;
  }
  
  export interface ICharacterDataContainer {
    offset: number;
    limit: number;
    total: number;
    count: number;
    results: Character[];
  }
  
  export interface ICharacter {
    id: number;
    name: string;
    description: string;
    modified: Date;
    resourceURI: string;
    urls: IUrl[];
    thumbnail: IImage;
    comics: IComicList;
    stories: IStoryList;
    events: IEventList;
    series: ISeriesList;
  }
  
  export interface ICharacterInput {
    data:ICharacter
  }
  export interface IUrl {
    type: string;
    url: string;
  }
  export interface IImage {
    path: string;
    extension: string;
  }
  export interface IComicList {
    available: number;
    returned: number;
    collectionURI: string;
    items: unknown[];
  }
  
  export interface IStoryList {
    available: number;
    returned: number;
    collectionURI: string;
    items: unknown[];
  }
  
  export interface IEventList {
    available: number;
    returned: number;
    collectionURI: string;
    items: unknown[];
  }
  
  export interface ISeriesList {
    available: number;
    returned: number;
    collectionURI: string;
    items: unknown[];
  }
  