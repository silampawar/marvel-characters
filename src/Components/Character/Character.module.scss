@import "../../Styles/theme.scss";

.container {
  width: 50%;
  max-width: 170px;
  position: relative;
  display: flex;
  align-items: flex-end;
  background-color: $color-black;
  margin: 10px 10px 10px 10px;
  @include for-mobile {
    width: 40%;
  }
  &:after {
    width: 0;
    height: 0;
    bottom: 0;
    right: 154px;
    content: "";
    position: absolute;
    border-style: solid;
    border-width: 16px 16px 0 0;
    border-color: red $color-white $color-white $color-white;
    @include for-mobile {
      right: 86.5%;
    }
  }

  @include for-tablet {
    width: 30%;
  }
}

.cover {
  color: $color-white;
  background-image: linear-gradient(to top , red, yellow);
  position: absolute;
  top: 0;
  left: 0;
  padding-bottom: 5px;
  width: 100%;
  height: 210px;
  z-index: 10;
  overflow: hidden;
  display: flex;
}

.summary {
  display: none;
  height: 100%;
  @include upper();

  li {
    list-style: none;
  }
}

.thumb {
  width: 100%;
  height: 210px;
  position: absolute;
  top: 0;
  left: 0;
  z-index: 20;
  background-color: $color-black;
  transition: opacity 0.3s linear;
}

.details {
  margin-top: 215px;
  padding: 20px;
  @include upper();
}

.title {
  font: $font-title;
  color: $color-white;
  transition: font-size 0.3s linear;
}

.descr {
  color: $color-grey;
}

.container:hover {
  .summary {
    display: block;
  }

  .thumb {
    opacity: 0;
  }

  .title {
    font-size: 17px;
  }
}