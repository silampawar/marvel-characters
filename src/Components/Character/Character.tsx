import React,{FC} from 'react';
import styles from "./Character.module.scss";
import {ICharacterInput} from "./Character.d"
const Character:FC<ICharacterInput> = ({data}:ICharacterInput):JSX.Element =>{

    const detailUrl = data && data.urls && data.urls.find((x: { type: string; }) => x.type === "detail");
    return <>
   <div className={styles.container}>
      <div className={styles.cover}>
        <ul className={styles.summary}>
          <li>
            <b>{data.comics.available}</b> comics
          </li>
          <li>
            <b>{data.events.available}</b> events
          </li>
          <li>
            <b>{data.series.available}</b> series
          </li>
          <li>
            <b>{data.stories.available}</b> stories
          </li>
        </ul>
      </div>
      <a href={detailUrl?.url} target="_blank" rel="noopener noreferrer">
        <img
          alt="thumb"
          className={styles.thumb}
          src={
            data.thumbnail?.path + "." + data.thumbnail?.extension
          }
        />
      </a>
      <div className={styles.details}>
        <div className={styles.detailsText}>
          <div className={styles.title}>{data.name}</div>
        </div>
      </div>
    </div>
    </>

}

export default Character;