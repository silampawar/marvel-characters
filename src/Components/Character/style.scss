.character {
    width: 200px;
    height: 300px;
    margin-bottom: 20px;
    overflow: hidden;
  
    button {
      background: none;
      border: none;
      outline: none;
      cursor: pointer;
    }
  
    figure {
      width: 168px;
      height: 282px;
  
      img {
        box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.4);
        margin-bottom: 10px;
      }
  
      figcaption {
        font: normal 16px "Open Sans", Arial;
        color: #333;
        text-align: center;
        text-shadow: 1px 1px 1px rgba(0, 0, 0, 0.4);
      }
    }
  }
  
  .character-detail {
    position: fixed;
    bottom: -500px;
    left: 0;
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-around;
    align-items: flex-start;
    width: 100%;
    height: 500px;
    background: #f50000;
    transition: all 0.3s ease;
  
    figure {
      img {
        box-shadow: none;
        border: 5px solid #fff;
        margin-bottom: 10px;
      }
  
      figcaption {
        font: normal 18px "Open Sans", Arial;
        color: #fff;
        text-shadow: none;
        text-align: left;
      }
    }
  }
  
  .character-detail-close-container {
    position: absolute;
    right: 120px;
    top: 20px;
    width: 0;
    height: 0;
  
    button {
      position: absolute;
      width: 100px;
      height: 40px;
      border: 1px solid #fff;
      background: none;
      color: #fff;
      box-shadow: none;
      border-radius: 4px;
      cursor: pointer;
    }
  }
  
  .character-detail--show {
    bottom: 0;
  }
  
  .character-detail-thumb {
    figure {
      width: 216px;
      height: 324px;
    }
  
    display: flex;
    justify-content: center;
    flex-grow: 1;
    padding: 25px 0 0;
    box-sizing: border-box;
  }
  
  .character-detail-content {
    display: flex;
    flex-grow: 3;
    flex-flow: column;
    justify-content: flex-start;
    align-items: center;
    height: 100%;
    padding: 25px 10px 10px;
    box-sizing: border-box;
    background: #000;
  
    h2,
    p {
      margin-bottom: 25px;
      color: #fff;
    }
  }
  
  @media screen and (min-width: 320px) {
    .character-detail-close-container {
      position: absolute;
      right: 120px;
      top: 20px;
      width: 0;
      height: 0;
  
      button {
        top: 420px;
        right: 10px;
        width: 100px;
        height: 40px;
        border: 1px solid #fff;
        background: none;
        color: #fff;
        box-shadow: none;
        border-radius: 4px;
        cursor: pointer;
      }
    }
  }