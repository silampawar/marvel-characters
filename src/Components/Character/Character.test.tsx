import React from "react";
import Character from "./Character";
import { shallow } from "enzyme";

const API_URL= 'https://gateway.marvel.com:443/v1/public/characters';

describe("Character", () => {
  test("should render properly", () => {
    const value: any = {
      urls: [{ type: "detail", url: API_URL }],
      comics: {},
      events: {},
      series: {},
      stories: {},
    };
    const wrapper = shallow(<Character data={value} />);
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find("li")).toHaveLength(4);
    
    expect(wrapper.find("a").props().href).toBe(API_URL);
  });
  test("Should render when no url present", () => {
    const value: any = {
      urls: [],
      comics: {},
      events: {},
      series: {},
      stories: {},
    };
    const wrapper = shallow(<Character data={value} />);

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find("a").props().href).toBeUndefined();
  });
});
