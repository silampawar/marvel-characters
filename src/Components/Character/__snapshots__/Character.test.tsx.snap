// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Should render when no url present 1`] = `ShallowWrapper {}`;

exports[`Character should have link without url if detail url not found 1`] = `ShallowWrapper {}`;

exports[`Character should render properly 1`] = `ShallowWrapper {}`;
